pkg/sql/schemachanger/scplan/internal/scgraph/dep_edge_tree.go

```diff
@@ -11,100 +11,217 @@
 package scgraph
 
 import (
+	"sync"
+
+	"github.com/cockroachdb/cockroach/pkg/sql/schemachanger/scpb"
 	"github.com/cockroachdb/cockroach/pkg/sql/schemachanger/screl"
 	"github.com/cockroachdb/cockroach/pkg/util/iterutil"
+	"github.com/cockroachdb/errors"
 	"github.com/google/btree"
 )
 
-type depEdgeTree struct {
-	t     *btree.BTree
-	order edgeTreeOrder
-	cmp   nodeCmpFn
-}
-
-type nodeCmpFn func(a, b *screl.Node) (less, eq bool)
-
-func newDepEdgeTree(order edgeTreeOrder, cmp nodeCmpFn) *depEdgeTree {
-	const degree = 8 // arbitrary
-	return &depEdgeTree{
-		t:     btree.New(degree),
-		order: order,
-		cmp:   cmp,
-	}
-}
-
 // edgeTreeOrder order in which the edge tree is sorted,
 // either based on from/to node indexes.
 type edgeTreeOrder bool
 
-func (o edgeTreeOrder) first(e Edge) *screl.Node {
-	if o == fromTo {
-		return e.From()
-	}
-	return e.To()
-}
-
-func (o edgeTreeOrder) second(e Edge) *screl.Node {
-	if o == toFrom {
-		return e.From()
-	}
-	return e.To()
-}
-
 const (
 	fromTo edgeTreeOrder = true
 	toFrom edgeTreeOrder = false
 )
 
-// edgeTreeEntry BTree items for tracking edges
-// in an ordered manner.
-type edgeTreeEntry struct {
-	t    *depEdgeTree
-	edge *DepEdge
+// getTargetIdxFunc is used to get the location of the node in the graph.
+type getTargetIdxFunc = func(n *screl.Node) targetIdx
+
+// depEdges is a data structure to store the set of depEdges. It offers
+// in-order traversal of edges from or to any node in the graph.
+type depEdges struct {
+	fromTo *btree.BTree
+	toFrom *btree.BTree
+
+	getTargetIdx getTargetIdxFunc
+	edgeAlloc    depEdgeAlloc
+	entryAlloc   depEdgeTreeEntryAlloc
+}
+
+// makeDepEdges constructs the depEdge structure.
+func makeDepEdges(getTargetIdx getTargetIdxFunc) depEdges {
+	const degree = 32 // arbitrary
+	return depEdges{
+		fromTo:       btree.New(degree),
+		toFrom:       btree.New(degree),
+		getTargetIdx: getTargetIdx,
+	}
 }
 
-func (et *depEdgeTree) insert(e *DepEdge) {
-	et.t.ReplaceOrInsert(&edgeTreeEntry{
-		t:    et,
-		edge: e,
+// insertOrUpdate will insert a new dep edge if no such edge exists between
+// from and to. Otherwise, it will update the edge accordingly. An error
+// may be returned if the kind is incompatible with the existing kind for
+// the edge.
+func (et *depEdges) insertOrUpdate(rule Rule, kind DepEdgeKind, from, to *screl.Node) error {
```

Comment:
In which cases would we update an rule? Does that imply we might have made a mistake in defining the rules in the first place?

Reply:
I added more commentary.